rules/prefer-string-starts-ends-with.js

```diff
@@ -64,40 +72,102 @@ const create = context => {
 				return;
 			}
 
-			context.report({
+			const [target] = node.arguments;
+			const method = result.messageId === MESSAGE_STARTS_WITH ? 'startsWith' : 'endsWith';
+
+			let isString = target.type === 'TemplateLiteral' ||
+				(
+					target.type === 'CallExpression' &&
+					target.callee.type === 'Identifier' &&
+					target.callee.name === 'String'
+				);
+			let isNonString = false;
+			if (!isString) {
+				const staticValue = getStaticValue(target, context.getScope());
+
+				if (staticValue) {
+					isString = typeof staticValue.value === 'string';
+					isNonString = !isString;
+				}
+			}
+
+			const problem = {
 				node,
-				messageId: result.messageId,
-				fix: fixer => {
-					const [target] = node.arguments;
-					const staticValue = getStaticValue(target, context.getScope());
-					if (staticValue && typeof staticValue.value !== 'string') {
-						// Ignore known, non-string value which won't have a startsWith/endsWith function.
-						return;
-					}
-
-					const method = result.messageId === MESSAGE_STARTS_WITH ? 'startsWith' : 'endsWith';
-					let targetString = sourceCode.getText(target);
-
-					if (
-						// If regex is parenthesized, we can use it, so we don't need add again
-						!isParenthesized(regexNode, sourceCode) &&
-						(isParenthesized(target, sourceCode) || shouldAddParenthesesToMemberExpressionObject(target, sourceCode))
-					) {
-						targetString = `(${targetString})`;
-					}
-
-					// The regex literal always starts with `/` or `(`, so we don't need check ASI
-
-					return [
-						// Replace regex with string
-						fixer.replaceText(regexNode, targetString),
-						// `.test` => `.startsWith` / `.endsWith`
-						fixer.replaceText(node.callee.property, method),
-						// Replace argument with result.string
-						fixer.replaceText(target, quoteString(result.string))
-					];
+				messageId: result.messageId
+			};
+
+			function * fix(fixer, fixType) {
+				let targetText = getParenthesizedText(target, sourceCode);
+				const isRegexParenthesized = isParenthesized(regexNode, sourceCode);
+				const isTargetParenthesized = isParenthesized(target, sourceCode);
+
+				switch (fixType) {
+					// Goal: `(target ?? '').startsWith(pattern)`
+					case FIX_TYPE_NULLISH_COALESCING:
+						if (
+							!isTargetParenthesized &&
+							shouldAddParenthesesToLogicalExpressionChild(target, {operator: '??', property: 'left'})
+						) {
+							targetText = addParentheses(targetText);
+						}
+
+						targetText += ' ?? \'\'';
+
+						// `LogicalExpression` need add parentheses to call `.startsWith()`,
+						// but if regex is parenthesized, we can reuse it
+						if (!isRegexParenthesized) {
+							targetText = addParentheses(targetText);
+						}
+
+						break;
+
+					// Goal: `String(target).startsWith(pattern)`
+					case FIX_TYPE_STRING_CASTING:
+						// `target` was a call argument, don't need check parentheses
+						targetText = `String(${targetText})`;
```

Comment:
This adds extra parenthesis that aren't needed sometimes. I think I had code to handle this before.

`(/^b/).test((a))`
`((a)).startsWith('b')`

Reply:
I know that, but better "add missing parens" instead of "remove extra parens".